### DISCUSSION TITLE
아이템 11. equals를 재정의하려거든 hashCode도 재정의하라

### DISCUSSION THREAD

**Discussion by JoisFe** [Study-2-Effective-Java/Effective-Java#23] [3. 과제] [answered]
# 아이템 11. equals를 재정의하려거든 hashCode도 재정의하라

이펙티브 자바 책에서 '고려해봐'가 아닌
### equals를 재정의한 클래스 모두에서 hashCode도 재정의해야 한다.!!
라고 못을 박고 시작한다. 그만큼 무조건 지켜야 하는 부분이다.

### 이유
equals를 재정의한 클래스에서 hashCode를 재정의하지 않을 시 hashCode 일반 규약을 어기게 되어 해당 클래스의 인스턴스를 HashMap이나 HashSet 같은 Collection 원소로 사용할 때 문제를 일으키키 때문
### Object 명세
- equals 비교에서 사용되는 정보가 변경되지 않았다면 애플리케이션이 실행되는 동안 그 객체의 hashCode 메서드는 몇 번을 호출해도 일관되게 항상 같은 값을 반환 해야 함 (단 애플리케이션을 다시 실행한다면 이 값이 달라져도 상관 없음)
- equals(Object)가 두 객체를 같다고 판단했다면 두 객체의 hashCode는 똑같은 값을 반환해야 함
- equals(Object)가 두 객체를 다르다고 판단했더라도 두 객체의 hashCode가 서로 다른 값을 반환할  필요는 없다. (단 다른 객체에 대해서는 다른 값을 반환해야 해시테이블의 성능이 좋아진다.)

위의 이유와 명세의 설명을 뒷받침 하기 위해 Java의 HashMap 자료구조에 대해 알아보고자 한다.

### HashMap이 해싱을 이용하여 어떻게 랜덤 읽기(V get(K))가 빠른 이유

HashMap의 경우 get(K) 경우 평균적으로 O(1)인 시간복잡도를 가진다. 그럼 어떻게 이렇게 빨리 해당 Key(k)에 매칭되는 Value를 찾을 수 있을까? Hash Map의 내부구조를 보면 배열로 이루어 져 있으며 Key가 배열의 인덱스가 될 수 있는데 이 것을 버킷이라 한다. Key가 어떤 원리로 내부 배열의 인덱스가 될 수 있게 되었는지 알아보기 위해 java에서 구현되어 있는 HashMap 클래스를 살펴보았다.

![public class HashMapK, V extends AbstractMapK, V](https://user-images.githubusercontent.com/90208100/209358022-8a86d300-1ac6-46c6-9ca9-8e01aeda57ef.png)

![Pasted Graphic 2](https://user-images.githubusercontent.com/90208100/209358042-1e2f78c0-b9f6-45bb-b51c-2d4d8f431ca1.png)

![static final int hash (Object key)](https://user-images.githubusercontent.com/90208100/209358058-7d0f18b1-7102-4040-b2b0-77d81fe71e71.png)

![Pasted Graphic 1](https://user-images.githubusercontent.com/90208100/209358075-7705ebef-1292-4a54-a880-f4c207148224.png)

![int n; K k;](https://user-images.githubusercontent.com/90208100/209358083-0ce2fc15-3d1b-4d04-a108-08966e32da55.png)

java.util 패키지에서 구현 되어있는 HashMap 클래스에서 get 메서드를 보면 key에 해당하는 value를 찾기 위해  getNode 메서드를 이용하여  해당 key의 <key value> 쌍의 node를  찾고 이때 hash(key) 값을 필요로 하는데 hash() 메서드에서 key.hashCode() 값을 이용하는 것을 확인할 수 있다. 해시함수의 인자를 key를 넣어 구한 해시값이 배열(Node<K, V>)의 인덱스가 되고 그 위치에 <key, value>가 저장되는 것이다.
 다시말해  해시함수 (hash() 메서드)의 인자에 key를 이용하여 얻은 해시값이 <key, value> 쌍을 저장한 배열의 인덱스를 구하고 Node<K, V> 배열의 해당 인덱스에 위치한  node를 찾아서 그 node의 value를 찾는 것이다. 따라서 key를 이용하면 hash 함수를 이용해 바로 배열에 자신이 위치한 index를 찾기 때문에 평균적으로 O(1)의 시간복잡도로 매우 빠르다.
하지만 getNode 메서드를 보면 해시함수에 키값을 넣어 얻은 hash 값으로 바로 원하는 <key, value>를 찾은것이 아니라 다른 복잡한 과정을 확인할 수 있다.
이 이유는 해시 충돌로 인한 것이다. 
### 해시 충돌이란 ? 
간단하게 말해서 해시 함수에 다른 키 값을 넣어도 같은 해시 값이 나올 수 있는데 이를 해시충돌이라 한다.
### 해시 충돌의 해결 방법
이러한 문제를 자바의 HashTable 에서는 Seperate Changing 방식으로 처리하는데  동일한 해시 값이 있을 경우 LinkedList로 관리하고 8개 이상인 경우 Tree로 변경하여 관리하는 방법을 이용한다.
![final int hash;](https://user-images.githubusercontent.com/90208100/209358103-b43becb1-90c3-407b-b3c1-05c7a54d90b2.png)

Node 클래스를 보면 Node<K, V>를 이용해 LinkedList로 구현되어 있음을 알 수 있다.
다시 getNode 메서드 일부를 보면 (충돌난 노드 갯수가 8개 이전에는 트리노드로 구현)
![return first;](https://user-images.githubusercontent.com/90208100/209358113-c4ebb897-9afb-4672-a4a8-0b6ff8996991.png)

만약 first.hash == hash 즉 배열에 저장된 node의 해시값(인덱스) 매개변수로 받아온 hash값이 같고 first.key == key 즉 배열에 저장된 node의 key와 매개변수로 받아온 key가 같아야 first를 반환한다. 해시값이 같으면서 key 값 까지 같아야 원하는 value를 찾는 것이다.
그리고 first .next != null 이 말은 2개 이상의 node가 연결리스트로 연결되어 있다는 의미이고 그것은 다른 key값이지만 해시함수의 결과값이 같은 해시 충돌이 일어난 값이 저장되어 있다는 뜻이된다.
따라서 이 경우 do while 문을 통해 해당 연결리스트를 순회하며 hash값과 key값 둘다 같은 node를 찾아야 하게되어 해당 value를 찾는데 시간이 더 걸리게 된다.

![overflow](https://user-images.githubusercontent.com/90208100/209358137-a01e89b5-f9da-47c7-81b3-0bb5e3b98796.png)

reference : https://commons.wikimedia.org/wiki/File:Hash_table_5_0_1_1_1_1_0_LL.svg
첨부한 이미지를 보면 keys John Smith와 Sandra Dee의 해시함수 결과가 152로 같았기 때문에 연결리스트로 두 (key, value) 쌍을 연결시킨 것을 확인할 수 있다.

### 정리
HashMap의 get(K) 메서드 경우 <key, value> 쌍에서 어떤 key에 해당하는 value를 찾기 위해서 해시 함수에 해당 key를 넣어 얻은 해시함수를 통해 <key, value>가 저장되어있는 배열에서의 해당 인덱스를 바로 구할 수 있어 평균 O(1) 시간 복잡도로 매우 빠르게 원하는 value를 찾을 수 있다. 그러나 해시 충돌로 인하여 같은 해시값을 가지는 키 값들 경우 그것들 끼리 연결리스트로 연결 되어 있어 그 연결리스트를 순회하며 원하는 value를 찾아야 하기 때문에 시간이 더 걸리게 된다. 그럼에도 불구하고 해시 충돌이 일어나는 <key, value>간의 연결리스트 들만 순회하면 되기 때문에 여전히 매우 빠르게 원하는 결과를 찾을 수 있다.

## HashMap에서 중복체크를 왜 못하지?
``` java
public final class PhoneNumber {

    private final short areaCode;
    private final short prefix;
    private final short lineNum;

    public PhoneNumber(short areaCode, short prefix, short lineNum) throws IllegalAccessException {
        this.areaCode = rangeCheck(areaCode, 999, "지역코드");
        this.prefix = rangeCheck(prefix, 999, "프리픽스");
        this.lineNum = rangeCheck(lineNum, 9999, "가입자 번호");
    }

    private static short rangeCheck(int val, int max, String arg) throws IllegalAccessException {
        if (val < 0 || val > max) {
            throw new IllegalAccessException(arg + ": " + val);
        }

        return (short) val;
    }
}
```

위의 코드는 equals, hashCode 모두 override를 하지 않은 클래스이다. 해당 클래스의 인스턴스를 HashMap의 키로 한다면 무슨일이 발생할까?

``` java
public final class PhoneNumber {

    private final short areaCode;
    private final short prefix;
    private final short lineNum;

    public PhoneNumber(short areaCode, short prefix, short lineNum) throws IllegalAccessException {
        this.areaCode = rangeCheck(areaCode, 999, "지역코드");
        this.prefix = rangeCheck(prefix, 999, "프리픽스");
        this.lineNum = rangeCheck(lineNum, 9999, "가입자 번호");
    }

    private static short rangeCheck(int val, int max, String arg) throws IllegalAccessException {
        if (val < 0 || val > max) {
            throw new IllegalAccessException(arg + ": " + val);
        }

        return (short) val;
    }
    
    @Override
    public String toString() {
        return "PhoneNumber{" +
            "areaCode=" + areaCode +
            ", prefix=" + prefix +
            ", lineNum=" + lineNum +
            '}';
    }
}
```
HashMap의 key에 PhoneNumber 객체 2개를 넣었다. 객체 인스턴스 변수가 모두 같게 하여 ker가 중복이 되어 map의 size가 1이 되길 희망했다.

![image](https://user-images.githubusercontent.com/90208100/209425702-b06c4500-807d-4dab-a6fe-af63e5706db1.png)

하지만 결과는 당연히 중복이라고 판단하지 않았고 size는 2가 나왔다.
주소값이 다를 테니 당연히 다른 객체로 판단할 것 이므로
Map은 중복으로 판단하지 않기 때문이다.

### 해결책 1 : equals를 재정의하여 해당 객체의 인스턴스 변수 값들이 모두 같으면 같은 객체로 보게 하자!

따라서 PhoneNumber 에서 equals를 재정의 하였다.

``` java
public final class PhoneNumber {

    private final short areaCode;
    private final short prefix;
    private final short lineNum;

    public PhoneNumber(short areaCode, short prefix, short lineNum) throws IllegalAccessException {
        this.areaCode = rangeCheck(areaCode, 999, "지역코드");
        this.prefix = rangeCheck(prefix, 999, "프리픽스");
        this.lineNum = rangeCheck(lineNum, 9999, "가입자 번호");
    }

    private static short rangeCheck(int val, int max, String arg) throws IllegalAccessException {
        if (val < 0 || val > max) {
            throw new IllegalAccessException(arg + ": " + val);
        }

        return (short) val;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        PhoneNumber that = (PhoneNumber) o;
        return areaCode == that.areaCode && prefix == that.prefix && lineNum == that.lineNum;
    }

    @Override
    public String toString() {
        return "PhoneNumber{" +
            "areaCode=" + areaCode +
            ", prefix=" + prefix +
            ", lineNum=" + lineNum +
            '}';
    }
}
```

이후 다시 결과를 보면

![image](https://user-images.githubusercontent.com/90208100/209425719-2947e828-cb46-44d1-9bc7-de5ea16a9e61.png)

여전히 map의 size가 2로 중복 체크가 되지 않았다.
왜 이럴까??
객체 주소는 다르지만 내부 인스턴스 값이 같다면 같은 객체로 보게끔 equals를 재정의 했음에도 중복체크가 안된다 ?? --> HashMap의 put 메서드에 대해서 자세히 보아야 할 것이다.

### 해결책2 : equals를 재정의 했으므로 hashCode도 재정의하라!!

equals 재정의 만으로는 중복체크가 안되었으니 HashMap의 put 메서드를 보자.
![image](https://user-images.githubusercontent.com/90208100/209425230-b9e274aa-3f39-4e03-ac2a-579f3a384b91.png)
put 메서드는 putVal 메서드를 이용하는데 단순히 key, value만 넘기는 것이 아니라 hash(key)로 key값의 hash code 값 또한 넘긴다.
putVal 메서드를 보면
![image](https://user-images.githubusercontent.com/90208100/209425287-3e469076-b1ad-4cb5-bf94-3ebc41b4fa23.png)
너무 복잡하다. 하지만 중복체크를 하는부분만 찾아보자

![image](https://user-images.githubusercontent.com/90208100/209425411-6819e084-6cd2-4170-aec3-cd894798dce9.png)

이전에 HashMap은 get을 할때 해시코드 값으로 위치를 찾아서 값을 가져왔다는 것을 설명하였다.
마찬가지로 put을 하기전에 해시코드 값으로 해당 위치에 갔더니 null ?? 당연히 아무것도 없는 것이니 중복자체가 생길 수 없다. 따라서 이때 map에 해당 k, v를 넣는다.

하지만 이미 존재한다면 ??

![image](https://user-images.githubusercontent.com/90208100/209425461-b5eec39d-203f-4a46-809a-392556799258.png)

여기서 중요하다 만약 해시 코드 값이 같고!!! (k = p.key) == key 여기서 equals를 쓰지 않았다. 즉 주소값 자체가 같다면 중복으로 보겠다는 것이다.
자 여기까지만 본다면 해시코드 값이 같다고 해서 중복으로 보지 않겠다는 것이다.

``` java
public final class PhoneNumber {

    private final short areaCode;
    private final short prefix;
    private final short lineNum;

    public PhoneNumber(short areaCode, short prefix, short lineNum) throws IllegalAccessException {
        this.areaCode = rangeCheck(areaCode, 999, "지역코드");
        this.prefix = rangeCheck(prefix, 999, "프리픽스");
        this.lineNum = rangeCheck(lineNum, 9999, "가입자 번호");
    }

    private static short rangeCheck(int val, int max, String arg) throws IllegalAccessException {
        if (val < 0 || val > max) {
            throw new IllegalAccessException(arg + ": " + val);
        }

        return (short) val;
    }
    
    @Override
    public int hashCode() {
        return Objects.hash(areaCode, prefix, lineNum);
    }

    @Override
    public String toString() {
        return "PhoneNumber{" +
            "areaCode=" + areaCode +
            ", prefix=" + prefix +
            ", lineNum=" + lineNum +
            '}';
    }
}
```

equals는 재정의하지 않고 hashCode만 재정의해도 결과는 
![image](https://user-images.githubusercontent.com/90208100/209425738-0739a17a-ed73-4ba3-9c04-074a2ce1ed6f.png)
여전히 size가 2인 이유다 (해시코드 값은 두 객체 모두 같다)

여기서 || (key != null && key.equals(k) 즉 key가 null이 아니고 equals메서드로 판단했을 경우 같다면 같다고 중복으로 판단하겠다는 의미가 된다.

### 정리
1. HashMap은 먼저 key값의 hash code 값으로 해당 key에 해당하는 value를 찾는다. 하지만 hashCode 메서드가 재정의 되어있지 않다면 key는 같지만 다른 hashCode 값을 가르키게 되어 value를 null을 찾게 되고 따라서 중복 체크를 하지 못한다.
2. HashMap에 put을 진행할 때 만약 hashCode 재정의가 되어있어도 key의 주솟값이 다르면 중복체크를 하지 않는다. 만약 이 경우 equals가 재정의 되어있어 같은 객체로 본다면 중복으로 본다.

따라서 equals, hashCode 메서드 모두 재정의해야만 제대로된 HashMap을 사용할 수 있게 된다. (중복 체크도 잘 되고)

결론으로 equals, hashCode 모두 재정의를 했을 경우
``` java
public final class PhoneNumber {

    private final short areaCode;
    private final short prefix;
    private final short lineNum;

    public PhoneNumber(short areaCode, short prefix, short lineNum) throws IllegalAccessException {
        this.areaCode = rangeCheck(areaCode, 999, "지역코드");
        this.prefix = rangeCheck(prefix, 999, "프리픽스");
        this.lineNum = rangeCheck(lineNum, 9999, "가입자 번호");
    }

    private static short rangeCheck(int val, int max, String arg) throws IllegalAccessException {
        if (val < 0 || val > max) {
            throw new IllegalAccessException(arg + ": " + val);
        }

        return (short) val;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        PhoneNumber that = (PhoneNumber) o;
        return areaCode == that.areaCode && prefix == that.prefix && lineNum == that.lineNum;
    }

    @Override
    public int hashCode() {
        return Objects.hash(areaCode, prefix, lineNum);
    }

    @Override
    public String toString() {
        return "PhoneNumber{" +
            "areaCode=" + areaCode +
            ", prefix=" + prefix +
            ", lineNum=" + lineNum +
            '}';
    }
}
```

결과 : 
![image](https://user-images.githubusercontent.com/90208100/209426161-e4d2c1e3-5ee7-466d-bdae-f81b64e496cc.png)

size를 1로 보는 것으로 보아 중복체크를 잘 한 것을 확인할 수 있다.

## hashCode 메서드를 어떻게 재정의 할 것인가 ?

``` java
@Override
    public int hashCode() {
        return Objects.hash(areaCode, prefix, lineNum);
    }
```

위의 예시에서 hashCode 메서드의 재정의를 이렇게 하였다. 왜 이렇게 만들었을까?

### 이상적인 해시 함수는 주어진 (서로 다른) 인스턴스들을 32비트 정수 범위에 균일하게 분배해야 한다.

이상적인 해시 함수를 완벽히 실현하기는 어렵지만 hashCode 메서드를 이상과 최대한 비슷하게 만드는 규칙이 존재한다.

1. int 변수 result를 선언한 후 값 c로 초기화 한다. 
> - c는 해당 객체의 첫번째 핵심 필드를 단게 2.a 방식으로 계산한 해시코드 (여기서 중요한 점은 핵심 필드란 equals 메서드에서 비교에 사용된 필드)
2. 해당 객체의 나머지 필드 f 각각에 대해 다음 작업을 수행한다.
  > a. 해당 필드의 해시코드 c를 계산한다.
  >   > (1). 기본 타입 필드라면 Type.hashCode(f)를 수행함 (여기서 Type은 해당 기본 타입의 박싱 클래스)
  >   > (2). 참조 타입 필드면서 이 클래스의 equals 메서드가 이 필드의 equals를 재귀적으로 호출해 비교한다면 이 필드의 hashCode를 재귀적으로 호출 함.
  >   >   > - 계산이 더 복잡해질 것 같으면 이 필드의 표준형(canonical representation)을 만들어 그 표준형의 hashCode를 호출한다.
  >   >   > - 필드의 값이 null이면 0을 사용한다 (다른 상수도 괜찮지만 전통적으로 0 사용)

  >   > (3). 필드가 배열이라면 핵심 원소 각각을 별도 필드처럼 다룬다. 이상의 규칙을 재귀적으로 적용해 각 핵심 원소의 해시코드를 계산한 다음, 단계 2.b 방식으로 갱신 한다.
  >   >   > - 배열에 핵심 원소가 하나도 없다면 단순히 상수(0을 추천)를 사용한다.
  >   >   > - 모든 원소가 핵심 원소라면 Arrays.hashCode를 사용한다.

  > b. 단계 2.a에서 계산한 해시코드 c로 result를 갱신한다.
3. result를 반환한다.

### equals 비교에 사용되지 않은 필드는 무조건, 반드시 hashCode 구현에 있어서 제외해야함 !!

``` java
public final class PhoneNumber {

    private final short areaCode;
    private final short prefix;
    private final short lineNum;

    public PhoneNumber(short areaCode, short prefix, short lineNum) throws IllegalAccessException {
        this.areaCode = rangeCheck(areaCode, 999, "지역코드");
        this.prefix = rangeCheck(prefix, 999, "프리픽스");
        this.lineNum = rangeCheck(lineNum, 9999, "가입자 번호");
    }

    private static short rangeCheck(int val, int max, String arg) throws IllegalAccessException {
        if (val < 0 || val > max) {
            throw new IllegalAccessException(arg + ": " + val);
        }

        return (short) val;
    }

  @Override
      public boolean equals(Object o) {
          if (this == o) {
              return true;
          }
          if (o == null || getClass() != o.getClass()) {
              return false;
          }
          PhoneNumber that = (PhoneNumber) o;
          return areaCode == that.areaCode && prefix == that.prefix && lineNum == that.lineNum;
      }
    
    @Override
    public String toString() {
        return "PhoneNumber{" +
            "areaCode=" + areaCode +
            ", prefix=" + prefix +
            ", lineNum=" + lineNum +
            '}';
    }
}
```

위 상황에서 hashCode 메서드를 만드는 규칙을 이용해서 hashCode 메서드를 재정의 해보면
``` java
    @Override
    public int hashCode() {
        int result = Short.hashCode(areaCode);
        
        result = 31 * result + Short.hashCode(prefix);
        result = 31 * result + Short.hashCode(lineNum);
        
        return result;
    }
```

위 hashCode 메서드는 위 규칙대로 인스턴스의 핵심 필드 3개만 사용한 간단한 계산만 수행한다.

해당 과정에서 비결정적(undetermined) 요소가 전혀 없기에 동치인 객체는 무조건 같은 해시코드 값을 가질 것!!

### 충분히 괜찮은 hashCode 메서드 이다.(단순하고, 충분히 빠르고, 서로 다른 전화번호들을 다른 해시 버킷들로 훌륭히 배분)

그러나 무언가 귀찮지 않나? 간단하게 하는 법이 없을까?
### Objects 클래스가 제공해주는 해시코드를 계산해주는 정적 메서드인 hash 메서드
Objects 클래스는 임의의 갯수 만큼 객체를 받아 해시코드를 계산해주는 정적 메서드인 hash를 제공

이 메서드 사용시 앞에서 규칙대로 구현한 hashCode 메서드와 비슷한 수준의 결과를 단 한 줄로 작성이 가능하다.

``` java
   @Override
    public int hashCode() {
        return Objects.hash(areaCode, prefix, lineNum);
    }
``` 

하지만 항상 편리한 것이 무조건 좋을 수는 없는 법!!

### 단점
Objects가 제공해주는 hash 메서드는 아쉽게도 속도가 느리다 (이전에 규칙대로 직접 작성한 hashCode 메서드에 비해)
- 입력 인수를 담기 위한 배열이 만들어 져야 하기 때문.
- 입력 중 기본 타입이 있다면 박싱과 언박싱 또한 거처야 하기 때문.

-> 따라서 hash 메서드는 성능에 민감하지 않은 상황에서만 사용하길 바람!!

### 캐싱
- 클래스가 불변이고 해시코드를 계산하는 비용이 크다면 매번 새로 계산하기 보다는 캐싱하는 방식을 고려해야함 !!
- 이 타입의 객체가 주로 해시의 키로 사용될 것 같다면 인스턴스가 만들어질 때 해시코드를 계산해둬야 함.

### Lazy Initialization (지연 초기화)
- 필드가 해시의 키로 사용되지 않는 경우 hashCode가 처음 불릴 때 계산하는 지연 초기화 전략을 사용할 수 있음
- 필드를 지연 초기화 하려면 해당 클래스를 thread safe 하게 만들도록 신경 써야 함 (아이템 83 참고)

PhoneNumber 클래스는 굳이 이렇게 할 이유는 없지만 예시를 위해 해보면

``` java
    private final short areaCode;
    private final short prefix;
    private final short lineNum;
    
    private int hashCode; // PhoneNumber에 핵심 필드가 아닌 필드가 존재한다면

  @Override
    public int hashCode() {
        int result = hashCode;
        
        if (result == 0) {
            result = Short.hashCode(areaCode);
            result = 31 * result + Short.hashCode(prefix);
            result = 31 * result + Short.hashCode(lineNum);
            hashCode = result;
        }
        
        return result;
    }
```

### 성능을 높이겠다고 해시코드를 계산할 때 핵심 필드를 생략해서는 안 된다.
- 속도는 빨라지겠지만 해시 품질이 나빠져 해시테이블의 성능을 심각하게 떨어트릴 가능성이 있음
- 특히 어떤 필드는 특정 영역에 몰린 인스턴스들의 해시코드를 넓은 범위로 퍼트려주는 효과가 있을지도 모름. (만약 이런 필드를 생략하면 해시테이블의 속도가 느려질 것임)

### hashCode가 반환하는 값의 생성 규칙을 API 사용자에게 자세히 공표하지 말아야 한다.
그래야만 클라이언트가 이 값에 의지하지 않게 되고 추후에 계산 방식을 바꿀 수 있게 된다.

## 정리
equals를 재정의 하거든 hashCode 또한 무조건 재정의 해라!!!!!!!!!!!!

**Comment by chikeem90** [comment]
@JoisFe 님

> HashMap은 먼저 key값의 hash code 값으로 해당 key에 해당하는 value를 찾는다. 하지만 hashCode 메서드가 재정의 되어있지 않다면 key는 같지만 다른 hashCode 값을 가르키게 되어 value를 null을 찾게 되고 따라서 중복 체크를 하지 못한다.
> HashMap에 put을 진행할 때 만약 hashCode 재정의가 되어있어도 key의 주솟값이 다르면 중복체크를 하지 않는다. 만약 이 경우 equals가 재정의 되어있어 같은 객체로 본다면 중복으로 본다.

👍 HashMap을 완전 분석(?) 하셨군요
HashXXX 컬렉션들은 내부적으로 hash function과 equals를 같이 쓰는 것 때문에 무조건 같이 재정의를 해줘야겠군요,,

**Comment by chikeem90** [comment]
equals랑 hashCode를 재정의할 때는,, 직접 구현하는 것도 좋지만 IDE의 도움을 받거나 lombok의 도움을 받으면 좋을 거 같네요.

1. intellij에서 생성해주는 equals / hashCode
   - 어떤 필드가 포함되고 제외될지 설정 가능
   - 내부적으로 Objects.hash를 사용
```java
public class Foo {
	int foo;
	String bar;

	@Override
	public boolean equals(Object o) {
		if (this == o)
			return true;
		if (o == null || getClass() != o.getClass())
			return false;
		Foo foo1 = (Foo)o;
		return foo == foo1.foo && Objects.equals(bar, foo1.bar);
	}

	@Override
	public int hashCode() {
		return Objects.hash(foo, bar);
	}
}
```

2. lombok에서 생성해주는 equals / hashCode
   - 책에 나온 방법과 거의 동일하게 hashCode를 직접 계산하도록 처리
```java
@EqualsAndHashCode
public class Foo {
	int foo;
	String bar;
}
```

```java
public class Foo {
    int foo;
    String bar;

    public Foo() {
    }

    public boolean equals(Object o) {
        if (o == this) {
            return true;
        } else if (!(o instanceof Foo)) {
            return false;
        } else {
            Foo other = (Foo)o;
            if (!other.canEqual(this)) {
                return false;
            } else if (this.foo != other.foo) {
                return false;
            } else {
                Object this$bar = this.bar;
                Object other$bar = other.bar;
                if (this$bar == null) {
                    if (other$bar != null) {
                        return false;
                    }
                } else if (!this$bar.equals(other$bar)) {
                    return false;
                }

                return true;
            }
        }
    }

    protected boolean canEqual(Object other) {
        return other instanceof Foo;
    }

    public int hashCode() {
        int PRIME = true;
        int result = 1;
        int result = result * 59 + this.foo;
        Object $bar = this.bar;
        result = result * 59 + ($bar == null ? 43 : $bar.hashCode());
        return result;
    }
}
```

요즘에는 lombok 사용이 거의 필수인데요,, 
equals와 hashCode 내부 동작은 책의 내용으로 이해하되, 필요하다면 lombok의 `@EqualsAndHashCode`를 잘 활용하면 좋을 거 같습니다.
`@EqualsAndHashCode.Exclude`, `@EqualsAndHashCode.Include`로 포함할지 말지 처리할 수 있네요.

https://projectlombok.org/features/EqualsAndHashCode

**Reply by JoisFe** [reply]
@EqualsAndHashCode.Exclude, @EqualsAndHashCode.Include 처리 방법 또한 존재했군요 감사합니다.!!!

lombok을 이용해 보면
``` java
import lombok.EqualsAndHashCode;

@EqualsAndHashCode
public class Test {

    @EqualsAndHashCode.Include
    private int num1;
    
    @EqualsAndHashCode.Exclude
    private int num2;
}
```
해당 코드를 delombok 해보니
``` java
public class Test {

    private int num1;

    private int num2;

    public boolean equals(final Object o) {
        if (o == this) {
            return true;
        }
        if (!(o instanceof Test)) {
            return false;
        }
        final Test other = (Test) o;
        if (!other.canEqual((Object) this)) {
            return false;
        }
        if (this.num1 != other.num1) {
            return false;
        }
        return true;
    }

    protected boolean canEqual(final Object other) {
        return other instanceof Test;
    }

    public int hashCode() {
        final int PRIME = 59;
        int result = 1;
        result = result * PRIME + this.num1;
        return result;
    }
}
```

시각적으로는 매우 많은 코드가 생략되어 편리하고 핵심필드와 핵심필드가 아닌 것 또한 잘 구분되니 사용한다면 유익할 것 같습니다.

**Comment by Irisation23** [comment, marked as answer]
`@EqualsAndHashCode.Exclude`, `@EqualsAndHashCode.Include` 저도 해당 어노테이션은 처음봐서 조금 더 찾아 봤습니다.

```java 
@EqualsAndHashCode(onlyExplicitlyIncluded = true)
public class Test {

    @EqualsAndHashCode.Include
    private int num1;
    private int num2;

    protected boolean canEqual(final Object other) {
        return other instanceof Test;
    }
}
```
@EqualsAndHashCode 해당 어노테이션의 default 값은 @EqualsAndHashCode(onlyExplicitlyIncluded = false) 입니다.
해당 경우는 클래스의 필드에 `@EqualsAndHashCode.Include`를 필드에 정의하지 않아도 핵심 필드로서 동작을 합니다.
하지만 위의 코드처럼 @EqualsAndHashCode(onlyExplicitlyIncluded = true) 값을 주게 된다면 반드시 핵심 필드를 `@EqualsAndHashCode.Include`로 정의 해주어야 한다는 사실을 알게 되었습니다.
Lombok이 굉장히 강력하네요. 
앞으로 Lombok 어노테이션 사용할 때 상태 값으로 줄 수있는 경우가 보인다면 조금 뜯어 보고 싶어졌습니다. 
감사합니다 👍

**Reply by chikeem90** [reply]
> @EqualsAndHashCode 해당 어노테이션의 default 값은 @EqualsAndHashCode(onlyExplicitlyIncluded = false) 입니다.

오,, 그렇군요 모든 필드에 애노테이션을 써야하는게 조금 번거롭다 느꼈는데,, 기본으로 그럼 include되고 뺄 놈만 exclude로 처리해두면 되겠네요,, 👍

**Reply by JoisFe** [reply]
좀더 구체적인 설명을 해주셔서 매우 감사합니다.
항상 새로운 개념을 알게 되면 그냥 사용하기 보다는 해당 부분에 대해 조금더 깊게 공부해보고 사용하는 것이 좋은 것 같습니다.👍
